### PR TITLE
[Trisha Patodia] iP 

### PR DESCRIPTION
**Pull request by triishaa**
# ChatterBox : Enhancing User Interaction 💬 

> "For every minute spent in organizing, an hour is earned." - Benjamin Franklin ([source](https://www.teachingcove.com/motivate/every-minute-spent-organizing/))

### Features Implemented: 
 - Add Tasks (`ToDo`, `Deadline`, and `Event`) 
 - Mark tasks as done ✅  or not done ❌ 
 - Save  and load tasks from file
 - Print tasks which all have a keyword 🔍 
 - Improve error handling for user inputs

 ### Next Steps:
1. **Refactor**  the code for better readability and maintainability.
2. **Optimise** file _I/O_ operations to enhance performance. 
3. ~~Add~~ **Revise** unit tests to cover edge cases and improve readability.

### Task List: 
- [x] Create a basic chatbot framework. 
- [x] Implement `add` , `delete` , `mark` and `search` functionalities.
- [ ] Improve user interface for better user experience. 

#### Here's a code snipped explaining how the `mark` functionality is implemented: 
```java 
private void markCommand(String input, TaskList taskList, Storage storage, Ui ui) {
        try {
            int indexSpace = input.indexOf(" ");
            int taskIndex = Integer.parseInt(input.substring(indexSpace + 1)) - 1;
            if (taskIndex >= 0 && taskIndex < taskList.size()) {
                taskList.getTask(taskIndex).markAsDone();
                storage.saveTasks(taskList.getTasks());
                ui.showTaskMarkedAsDone(taskList.getTask(taskIndex));
            } else {
                ui.showErrorInvalidTaskNumber();
            }
        } catch (NumberFormatException e) {
            ui.showError("Command must be followed by a specific task number");
        }
    }
```

### CODE REVIEW

File: src/main/java/command/Parser.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    /**
+     * Parses and executes user command based on the input string.
+     *
+     * @param input The command input from the user.
+     * @param taskList The list of tasks managed by the chatbot.
+     * @param storage The storage handler responsible for saving tasks.
+     * @param ui The user interface handler for displaying messages.
+     */
+    public void parseExecute(String input, TaskList taskList, Storage storage, Ui ui) {
+        if(input.startsWith("mark")) {
```
Comment (k-hian) [inline src/main/java/command/Parser.java]: Take note on single space within the statements. Here, there should be a space between `if` and the condition.

File: src/main/java/task/Deadline.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represents a task with a deadline.
+ * A Deadline task has a description and a specific date and time
+ * by which it needs to be completed.
+ */
+public class Deadline extends Task{
```
Comment (k-hian) [inline src/main/java/task/Deadline.java]: Take note of the missing single space between `Task` and `{`.

Similarly for a few other classes as well.

File: src/test/java/command/UiTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import java.io.PrintStream;
+import java.util.ArrayList;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class UiTest {
+    private Ui ui;
+    private final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    private final PrintStream originalOut = System.out;
+
+    @BeforeEach
```
Comment (k-hian) [inline src/test/java/command/UiTest.java:18]: First time seeing a setup and teardown method. Learned something new :+1:

**Review by k-hian — COMMENTED** [review]
LGTM. Just some minor nits to fix.

File: src/main/java/task/Deadline.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        this.by = by;
+    }
+
+    /**
+     * Returns the description of the deadline task.
+     *
+     * @return The description of the deadline task.
+     */
+    @Override
+    public String getDescription() {
```
Comment (nihalramesh) [inline src/main/java/task/Deadline.java:43]: Perhaps we could get rid of this method, as it adds no functionality as compared to its parent method?

File: src/main/java/task/Event.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+     *
+     * @return The LocalDateTime object representing the end date and time of the event.
+     */
+    public LocalDateTime getEndDate() {
+        return this.dateEnd;
+    }
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: "
+                + dateStart.format(DateTimeFormatter.ofPattern("MMM dd yyyy hh:mm a"))
```
Comment (nihalramesh) [inline src/main/java/task/Event.java:76]: To avoid repeated code and ensure consistency, consider defining a DateTimeFormatter as a private static final constant for the two formats used in toString() and toSaveFormat().

File: src/main/java/command/Parser.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+    /**
+     * Parses and executes user command based on the input string.
+     *
+     * @param input The command input from the user.
+     * @param taskList The list of tasks managed by the chatbot.
+     * @param storage The storage handler responsible for saving tasks.
+     * @param ui The user interface handler for displaying messages.
+     */
+    public void parseExecute(String input, TaskList taskList, Storage storage, Ui ui) {
```
Comment (nihalramesh) [inline src/main/java/command/Parser.java]: Perhaps we could rename this to something more descriptive?

File: src/main/java/command/Parser.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+     * Adds a Deadline task based on the user command.
+     *
+     * @param input The command input specifying the Deadline task.
+     * @param taskList The list of tasks managed by the chatbot.
+     * @param storage The storage handler responsible for saving tasks.
+     * @param ui The user interface handler for displaying messages.
+     */
+    private void deadlineCommand(String input, TaskList taskList, Storage storage, Ui ui) {
+        if (!input.contains("/by")) {
+            ui.showError("task.Deadline format should be: deadline DESCRIPTION /by DATE");
```
Comment (nihalramesh) [inline src/main/java/command/Parser.java]: Perhaps we could standardised error messages?

**Review by nihalramesh — COMMENTED** [review]
Overall LGTM, just a bit of nit picking here and there. Cheers!
